src/Features/Core/Portable/ConvertForToForEach/AbstractConvertForToForEachCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,459 @@
+
+            // See if the first statement in the for loop is of the form:
+            //      var x = list[i]   or
+            //
+            // If so, we'll use those as the iteration variables for the new foreach statement.
+            var bodyStatements = GetBodyStatements(forStatement);
+            var (typeNode, foreachIdentifier, declarationStatement) = tryDeconstructInitialDeclaration();
+
+            if (typeNode == null)
+            {
+                // user didn't provide an explicit type.  Check if the index-type of hte collection
+                // is different from than .Current type of the enumerator.  If so, add an explicit
+                // type so that the foreach will coerce the types accordingly.
+                var indexerType = GetIndexerType(containingType, collectionType);
+                if (!Equals(indexerType, iterationType))
+                {
+                    typeNode = (TTypeNode)generator.TypeExpression(
+                        indexerType ?? semanticModel.Compilation.GetSpecialType(SpecialType.System_Object));
+                }
+            }
+
+            // If we couldn't find an appropriate existing variable to use as the foreach
+            // variable, then generate one automatically.
+            if (foreachIdentifier.RawKind == 0)
+            {
+                foreachIdentifier = semanticFacts.GenerateUniqueName(
+                    semanticModel, forStatement, containerOpt: null, baseName: "v", cancellationToken);
+                foreachIdentifier = foreachIdentifier.WithAdditionalAnnotations(RenameAnnotation.Create());
+            }
+
+            // Create the expression we'll use to replace all matches in the for-body.
+            var foreachIdentifierReference = foreachIdentifier.WithoutAnnotations(RenameAnnotation.Kind).WithoutTrivia();
+
+            // Walk the for statement, replacing any matches we find.
+            findAndReplaceMatches(forStatement);
+
+            // Finally, remove the declaration statement if we found one.  Move all its leading
+            // trivia to the next statement.
+            if (declarationStatement != null)
+            {
+                editor.RemoveNode(declarationStatement,
+                    SyntaxGenerator.DefaultRemoveOptions | SyntaxRemoveOptions.KeepLeadingTrivia);
+            }
+
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            editor.ReplaceNode(
+                forStatement,
+                (currentFor, _) => this.ConvertForNode(
+                    (TForStatementSyntax)currentFor, typeNode, foreachIdentifier,
+                    collectionExpression, iterationType, options));
+
+            return document.WithSyntaxRoot(editor.GetChangedRoot());
+
+            // local functions
+            (TTypeNode, SyntaxToken, TStatementSyntax) tryDeconstructInitialDeclaration()
+            {
+                if (bodyStatements.Count >= 1)
```

Comment:
Consider inverting those conditions (bail if condition isn't met)

Reply:
in this case i prefer.  i know though that the deep-nesting-of-doom isn't great :-/

Reply:
interesting. so, rather than explicitly getting inputs, it uses captures. won't it be dangerous since someone without realizing the variable is captured and used later, add or replace the variable?

Reply:
bodyStatements is actually not used anywhere else except this method, it could brought down to this method.

Reply:
> bodyStatements is actually not used anywhere else except this method, it could brought down to this method.

Sure!

Reply:
#resolved